File: printer_server/drivers/hexapod/hexapod_snip.py
import printer_server.views.manual_controls
from printer_server.extensions import socketio
from printer_server.hardware_configuration.hardware_configuration import driver_handles

hexapod = driver_handles.hexapod

PIVOT_SET_COMMAND = "pivot"
TRANSLATION_COMMAND = "translation"
ROTATION_COMMAND = "rotation"

@socketio.on("hexapod_get_positions", namespace="/manual")
def get_hexapod_positions(emit=True, log=False):
    last_positions = (
        printer_server.views.manual_controls.get_last_calibration_positions_from_logs()
    )

    new_tip = hexapod.get_pose("Tip")*1000
    new_tilt = hexapod.get_pose("Tilt")*1000
    new_rotate = hexapod.get_pose("Rotate")*1000

    if new_tip is not None:
        last_positions["tip"] = new_tip
        last_positions["tilt"] = new_tilt
        last_positions["rotate"] = new_rotate

    if log:
        printer_server.views.manual_controls.write_to_position_log(last_positions)

    if emit:
        socketio.emit(
            "hexapod_done",
            last_positions,
            namespace="/manual"
        )

    return last_positions

@socketio.on("initialize_hexapod", namespace="/manual")
def initialize_hexapod():
    init_flag = hexapod.initialized
    if (init_flag):
        error_codes = hexapod.get_status()
        pivot_update = hexapod.get_pivot_point()
        pose_update = hexapod.get_pose()
        
    else:
        error_codes = None
        pivot_update = None
        pose_update = None

    socketio.emit("init_msg", [init_flag, error_codes, pivot_update, pose_update], namespace="/manual")

@socketio.on("check_init", namespace="/manual")
def check_init():
    init_flag = hexapod.initialized
    if (init_flag):
        error_codes = hexapod.get_status()
        pivot_update = hexapod.get_pivot_point()
        pose_update = hexapod.get_pose()
        
    else:
        error_codes = None
        pivot_update = None
        pose_update = None

    socketio.emit("init_msg", [init_flag, error_codes, pivot_update, pose_update], namespace="/manual")

@socketio.on("stop_motion", namespace="/manual")
def stop_motion():
    hexapod.hard_stop()
    pose_update = hexapod.get_pose()

    socketio.emit("pose_update", pose_update, namespace="/manual")

@socketio.on("axis_step", namespace="/manual")
def axis_step(step_information):
    command_type = step_information[0]
    axis = step_information[1]
    value = step_information[2]

    # Perform hexapod command
    if (command_type == PIVOT_SET_COMMAND):
        hexapod.step_pivot_point(axis, value)
        pivot_update = hexapod.get_pivot_point()
        socketio.emit("pivot_update", pivot_update, namespace="/manual")

    elif (command_type == TRANSLATION_COMMAND) or (command_type == ROTATION_COMMAND):
        hexapod.step_axis(axis, value)
        pose_update = hexapod.get_pose()
        socketio.emit("pose_update", pose_update, namespace="/manual")    
        get_hexapod_positions(log=True)

@socketio.on("pivot_command", namespace="/manual")
def pivot_command(pivot_information):
    x_pivot_pos = pivot_information[0]
    y_pivot_pos = pivot_information[1]
    z_pivot_pos = pivot_information[2]

    hexapod.set_pivot_point(x_pivot_pos, y_pivot_pos, z_pivot_pos)
    pivot_update = hexapod.get_pivot_point()

    socketio.emit("pivot_update", pivot_update, namespace="/manual")

@socketio.on("pose_command", namespace="/manual")
def pose_command(pose_information):
    x = pose_information[0]
    y = pose_information[1]
    z = pose_information[2]
    u = pose_information[3]
    v = pose_information[4]
    w = pose_information[5]

    hexapod.move_to_position_compound(x, y, z)
    hexapod.move_to_angle_compound(u, v, w)
    # hexapod.set_pose(x, y, z, u, v, w)
    pose_update = hexapod.get_pose()

    socketio.emit("pose_update", pose_update, namespace="/manual")
    get_hexapod_positions(log=True)

@socketio.on("request_dynamic_ranges", namespace="/manual")
def request_dynamic_ranges():
    ranges = {
        "X": hexapod.get_simple_dynamic_range('X'),
        "Y": hexapod.get_simple_dynamic_range('Y'),
        "Z": hexapod.get_simple_dynamic_range('Z'),
        "U": hexapod.get_simple_dynamic_range('U'),
        "V": hexapod.get_simple_dynamic_range('V'),
        "W": hexapod.get_simple_dynamic_range('W')
    }
    socketio.emit("dynamic_ranges", ranges, namespace="/manual")

File: printer_server/drivers/hexapod/hexapod_snip.js
$(document).ready(function () {

    check_initialization();

    // ====================== HELPER METHODS
    function send_event(event_name, param_list = []) {
        console.log(event_name);
        if (param_list.length == 0) {
            socket.emit(event_name);
        }
        else {
            console.log(param_list);
            socket.emit(event_name, param_list);
        }
    }

    function update_pivot_fields(new_pivot) {

        // Update html elements
        document.getElementById("pivot_R_state").innerHTML = new_pivot.R.toFixed(0) + "um";
        document.getElementById("pivot_S_state").innerHTML = new_pivot.S.toFixed(0) + "um";
        document.getElementById("pivot_T_state").innerHTML = new_pivot.T.toFixed(0) + "um";

        document.getElementById("pivot_R_state_input").value = parseFloat(new_pivot.R.toFixed(0));
        document.getElementById("pivot_S_state_input").value = parseFloat(new_pivot.S.toFixed(0));
        document.getElementById("pivot_T_state_input").value = parseFloat(new_pivot.T.toFixed(0));
    }

    function update_rotation_fields(rotation) {
        console.log("rotation update", rotation);

        // Update html elements
        document.getElementById("rotation_U_state").innerHTML = (rotation[0]).toFixed(1) + " mrad";
        document.getElementById("rotation_V_state").innerHTML = (rotation[1]).toFixed(1) + " mrad";
        document.getElementById("rotation_W_state").innerHTML = (rotation[2]).toFixed(1) + " mrad";

        document.getElementById("rotation_U_state_input").value = parseFloat((rotation[0]).toFixed(1));
        document.getElementById("rotation_V_state_input").value = parseFloat((rotation[1]).toFixed(1));
        document.getElementById("rotation_W_state_input").value = parseFloat((rotation[2]).toFixed(1));
    }

    function update_translation_fields(translation) {
        console.log("translation update", translation);

        // Update html elements
        document.getElementById("translation_X_state").innerHTML = (translation[0]).toFixed(0) + "um";
        document.getElementById("translation_Y_state").innerHTML = (translation[1]).toFixed(0) + "um";
        document.getElementById("translation_Z_state").innerHTML = (translation[2]).toFixed(0) + "um";

        document.getElementById("translation_X_state_input").value = parseFloat((translation[0]).toFixed(0));
        document.getElementById("translation_Y_state_input").value = parseFloat((translation[1]).toFixed(0));
        document.getElementById("translation_Z_state_input").value = parseFloat((translation[2]).toFixed(0));
    }

    function check_initialization() {
        send_event("check_init");
    }

    $(`.hexapod-cntrl-btn`).click(function () {
        let command_elements = $(this).attr("id").split("_");

        let command_type = command_elements[0];
        let axis = command_elements[1];
        let value = parseFloat(command_elements[2]);

        console.log("command_type", command_type);
        console.log("axis", axis);
        console.log("value", value);

        send_event("axis_step", [command_type, axis, value]);
    });

    var init_btn = document.getElementById("init_btn");
    var set_pivot_btn = document.getElementById("set_pivot_btn");
    var send_command_btn = document.getElementById("send_command_btn");
    var stop_motion_btn = document.getElementById("stop_motion_btn");

    init_btn.onclick = function () {
        send_event("initialize_hexapod");
    }

    socket.on("dynamic_ranges", function (ranges) {
        console.log(ranges);
    });

    socket.on("init_msg", function (init_output) {
        let init_flag = init_output[0]
        let init_error_codes = init_output[1];
        let init_pivot = init_output[2];
        let init_pose = init_output[3];

        let rotation = [init_pose["U"], init_pose["V"], init_pose["W"]];
        let translation = [init_pose["X"], init_pose["Y"], init_pose["Z"]];

        if (init_flag) {
            // Update fields
            update_pivot_fields(init_pivot);
            update_translation_fields(translation);
            update_rotation_fields(rotation);
        }
    });

    set_pivot_btn.onclick = function () {
        // Compile inputs from all the boxes
        let pivot_R_state_text = parseFloat(document.getElementById("pivot_R_state_input").value);
        let pivot_S_state_text = parseFloat(document.getElementById("pivot_S_state_input").value);
        let pivot_T_state_text = parseFloat(document.getElementById("pivot_T_state_input").value);

        // console.log("pivot_R_state_text", pivot_R_state_text);
        // console.log("pivot_S_state_text", pivot_S_state_text);
        // console.log("pivot_T_state_text", pivot_T_state_text);

        // Pack values on a list
        let new_pivot_point = [pivot_R_state_text, pivot_S_state_text, pivot_T_state_text];

        // Send event
        send_event("pivot_command", new_pivot_point);
    }

    socket.on("pivot_update", function (pivot) {
        console.log("pivot update", pivot);
        update_pivot_fields(pivot);
    });

    send_command_btn.onclick = function () {
        // Compile inputs from all the boxes
        let translation_X_state_text = parseFloat(document.getElementById("translation_X_state_input").value);
        let translation_Y_state_text = parseFloat(document.getElementById("translation_Y_state_input").value);
        let translation_Z_state_text = parseFloat(document.getElementById("translation_Z_state_input").value);
        let rotation_U_state_text = parseFloat(document.getElementById("rotation_U_state_input").value);
        let rotation_V_state_text = parseFloat(document.getElementById("rotation_V_state_input").value);
        let rotation_W_state_text = parseFloat(document.getElementById("rotation_W_state_input").value);


        // console.log("translation_X_state_text", translation_X_state_text);
        // console.log("translation_Y_state_text", translation_Y_state_text);
        // console.log("translation_Z_state_text", translation_Z_state_text);
        // console.log("rotation_U_state_text", rotation_U_state_text);
        // console.log("rotation_V_state_text", rotation_V_state_text);
        // console.log("rotation_W_state_text", rotation_W_state_text);

        // Pack values on a list
        let new_pose = [translation_X_state_text, translation_Y_state_text, translation_Z_state_text,
            rotation_U_state_text, rotation_V_state_text, rotation_W_state_text]

        // Send event
        send_event("pose_command", new_pose);
    }

    socket.on("pose_update", function (pose) {
        console.log("pose update", pose);

        let rotation = [pose["U"], pose["V"], pose["W"]];
        let translation = [pose["X"], pose["Y"], pose["Z"]];

        update_translation_fields(translation);
        update_rotation_fields(rotation);
    });

    stop_motion_btn.onclick = function () {
        send_event("stop_motion");
    }

    var request_dynamic_ranges_btn = document.getElementById("request_dynamic_ranges");
    request_dynamic_ranges_btn.onclick = function () {
        send_event("request_dynamic_ranges");
    }
});